### DISCUSSION TITLE
How to add edit the CronSchedule with python script (instead of webui)?

### DISCUSSION THREAD

**Discussion by mr-tester-william-lo** [PrefectHQ/prefect#4628] [General] [answered]
Currently, i have this code:

```
from prefect import Flow
from prefect.schedules import CronSchedule
from prefect.tasks.prefect import StartFlowRun
from datetime import datetime

weekday_schedule = CronSchedule(
    "30 * * * 2-3", start_date=datetime.now()
)

flow_1 = StartFlowRun(
                flow_name="this_is_a_flowname", 
                project_name="this_is_a_projname",
                parameters={ "Task1_Var": "11111", "Task2_Var": "22222" },
                wait=True
                )
with Flow("parent-flow", schedule=weekday_schedule) as flow:
    flow_1
flow.register( project_name = "testing" )
```

This workflow has been registered in prefect server/cloud. I found the web-UI can edit the schedule, but i cannot find any related package in python which allows me to add/remove/edit the CronSchedule or change to another schedule package "IntervalSchedule".  So, how can I do it in python? 

Note that I prefer I can set my schedule out of my flow code with the use of python.

**Comment by kvnkho** [comment, marked as answer]
Hi @mr-tester-william-lo ,

I am not understanding 100% so I'll just try to answer as best I can.

First, I just want to make sure you know that you can combine `CronClocks` and `IntervalClocks` to make a more complex schedule. You can do something like:
```python
clock1   = clocks.IntervalClock(...)
clock2   = clocks.CronClock(...)

# the full schedule
schedule = Schedule(clocks=[clock1, clock2])
```

From my understanding, your Flow has a `CronSchedule` and you want to change to `IntervalSchedule`. All you need to do is change your weekday_schedule to the `IntervalSchedule` and re-register the flow for it to take effect.

```python
schedule = IntervalSchedule(interval=timedelta(minutes=2))

with Flow("parent-flow", schedule=schedule) as flow:
    flow_1
flow.register( project_name = "testing" )
```
This will replace the schedule for you.

Now on the last bit, you want to set the schedule outside of Flow code. You can do something like this from another file:
```python
from prefect.utilities.storage import extract_flow_from_file

def register_flow(file, project, labels):
    file_path = os.path.abspath(file)
    flow = extract_flow_from_file(file_path=file_path)
    flow.schedule = IntervalSchedule(interval=timedelta(minutes=2))
    flow.register(project_name=project, labels=labels, no_url=True)
```

I don't recommend this though because setting Executor and Results this way will not work. Those need to be defined in the script where you have the Flow. This will work for schedules though. And then as you mentioned, the UI is also another way to edit schedules of registered flows.

What is your use case that you need the schedule detached from the Flow code?

**Reply by mr-tester-william-lo** [reply]
Thanks so much for your answer, let me try to explain a bit more.

I made a generic test flow (Flow A) to get files from folders and try to visualize them in log in different time period.
Thus, I added the parameter in generic flow (Flow A) which allow me to point to different path ID and to get different files data.
![Screenshot 2021-06-09 at 9 41 43 AM](https://user-images.githubusercontent.com/83815028/121280212-7c769680-c908-11eb-8089-4a116d825cda.png)


Next, I try to check out how to add a schedule in "**Existing** Flow". Unfortunately, other than building parent flow, I find nothing related to build a schedule to utilize "**Existing** Flow".

So, here I have 2 questions, related to how to implement in python

- how to add multiple schedules to "**Existing** Flow" with multiple parameters?
- what should I do to detach the existing default schedule without register again? (note, the flow is stored in docker)

If WebUI makes good use of GQL as API interface to create and edit current schedule with parameters. Suppose i can do it in Python coding as well. So, what should i do to make it happen? May i know if there is any related document mentioned that. As there are thousand schedulers would like to add, suppose there are some ways to create schedules instead of do it manually

**Reply by kvnkho** [reply]
Sorry for the slight delay. 

I get what you're saying now. Yes you can use the GraphQL API to set schedules on Flows. There is a `set_flow_group_schedule` that you can use like this:

```python
from prefect.client.client import Client

client = Client()

query = """
mutation {
  set_flow_group_schedule (input: {
      flow_group_id: "795512b7-c9f7-4231-8d6b-c80dbd87231d", 
      cron_clocks: [{cron: "0 9 * * *", parameter_defaults: "{\\"x\\": 1}"}]}) {
    success
    error
  }
}
"""
client.graphql(query)

```

Note that `cron_clocks` takes in multiple clocks here with multiple parameters. This mutation also takes in interval clocks, you can find more details in the Interactive API tab of the UI. To delete existing schedules from Flows is a similar mutation.

```python
mutation {
  delete_flow_group_schedule (input: {flow_group_id: "d3a03515-8876-4564-aa45-382b394110a3"}) {
    success
    error
  }
}
```

**Reply by mr-tester-william-lo** [reply]
Thanks, problem solved. Not much information in the documentation. 
Below link find more Graphql Query Method
https://github.com/PrefectHQ/prefect/tree/master/docs/orchestration/concepts

**Reply by mr-tester-william-lo** [reply]
May I know how to add labels to the schedule created by graphql?

i can only find this:
https://github.com/PrefectHQ/prefect/blob/d61fa6aac9330c5817cc8e8b8f8cca2d634ea7e1/src/prefect/schedules/clocks.py#L255
, but it is not related to graphql query

**Reply by zhen0** [reply]
Hi @mr-tester-william-lo - thanks for the question.  

Labels aren't currently exposed as an input for the `set_flow_group_schedule` graphQL mutation.  The code you are referencing is for setting a flow in code (where you can add labels as part of a cron_clock).  

You can add labels separately using the `set_flow_group_labels` mutation but that will affect all flow runs, not just the ones created by a certain schedule. 

If you'd like to be able to set labels for a schedule through the graphQL API I suggest you open an issue so we can look into it. 

Also if you need documentation for what you can do using the graphQL API, take a look at the Interactive API in the web UI. It has a docs section listing the graphQL API schema.

**Reply by mr-tester-william-lo** [reply]
if 

- I have 2 agents, one is agent1 from cluster1 and another one is agent2 from cluster 2,
- They share the same code in prefect workflow but different parameters, 
- the prefect workflow label as (label1, label2), agent1 label as (label1) and agent2 label as (label2)

Then, do you mean:

- i have no choice to schedule and run them in the corresponding cluster if I do not **hard code** the schedule in the workflow?

Note that  in above test case, both agents will be triggered by the schedule run.
